File: LAB2_Ibaca_Lopez.java
import java.io.*;
import java.util.*;
import java.util.concurrent.*;

public class LAB2_Ibaca_Lopez {

	// =============================
	// Clases auxiliares
	// =============================

	static class Arista {
		int origen, destino, peso;
		Arista(int o, int d, int p) {
			origen = o; 
            destino = d; 
            peso = p;
		}
	}

	static class Actualizacion {
		int nodo, distancia, padre;
		Actualizacion(int n, int d, int p) {
			nodo = n; 
            distancia = d; 
            padre = p;
		}
	}

	// =============================
	// Variables globales
	// =============================

	static int numHilos;
	static int numNodos;
	static int nodoInicio;
	static int nodoFinal;
	static ArrayList<Arista> aristas = new ArrayList<>();
	static int[] distancias;
	static int[] predecesores;

	// =============================
	// Métodos principales
	// =============================

	// Lee el archivo de entrada y carga los datos
	static void leerArchivo(String nombreArchivo) {

		try {
			BufferedReader br = new BufferedReader(new FileReader(nombreArchivo));
			String linea;

			// lee los primeros datos
			numHilos = Integer.parseInt(br.readLine().trim());
			br.readLine(); // Saltar numProcesos (solo para C/C++)
			numNodos = Integer.parseInt(br.readLine().trim());
			nodoInicio = Integer.parseInt(br.readLine().trim());
			nodoFinal = Integer.parseInt(br.readLine().trim());

			aristas.clear();
			while ((linea = br.readLine()) != null) {
				linea = linea.trim();
				if (linea.isEmpty()) continue;
				String[] partes = linea.split("\\s+");
				if (partes.length == 3) {
					int origen = Integer.parseInt(partes[0]);
					int destino = Integer.parseInt(partes[1]);
					int peso = Integer.parseInt(partes[2]);
					aristas.add(new Arista(origen, destino, peso));
				}
			}
			br.close();

		} catch (IOException | NumberFormatException e) {
			System.err.println("Error al leer el archivo: " + e.getMessage());
			System.exit(1);
		}
	}

	// Inicializa distancias y predecesores en infinito (excepto el nodo de inicio)
	static void inicializarBellmanFord() {
		
		distancias = new int[numNodos];
		predecesores = new int[numNodos];

		for (int i = 0; i < numNodos; i++) {
			distancias[i] = Integer.MAX_VALUE; // Representa infinito
			predecesores[i] = -1; // Sin predecesor
		}

		distancias[nodoInicio] = 0;// la distancia al nodo de inicio es 0
	}

	// Divide la lista de aristas en sublistas, asignando las aristas de forma equitativa (round-robin) a 
	// cada hilo para balancear la carga de trabajo.
	static ArrayList<ArrayList<Arista>> dividirAristas() {

		ArrayList<ArrayList<Arista>> particiones = new ArrayList<>();

		for (int i = 0; i < numHilos; i++) {// inicializa las sublistas vacías
			particiones.add(new ArrayList<>());
		}

		for (int i = 0; i < aristas.size(); i++) {// distribuye las aristas de manera equitativa (round-robin)
			int indiceHilo = i % numHilos;
			particiones.get(indiceHilo).add(aristas.get(i));
		}

		return particiones;
	}

	// Cada hilo procesa un subconjunto de aristas y, si encuentra una relajación posible,
	// propone una actualización de distancia y predecesor agregándola a una lista compartida.
	static class HiloBellmanFord extends Thread {
		ArrayList<Arista> aristasAsignadas;
		List<Actualizacion> actualizacionesCompartidas;

		// Constructor recibe las aristas y la lista compartida de actualizaciones
		HiloBellmanFord(ArrayList<Arista> aristas, List<Actualizacion> actualizacionesCompartidas) {
			this.aristasAsignadas = aristas;
			this.actualizacionesCompartidas = actualizacionesCompartidas;
		}

		/**
		 * Procesa las aristas asignadas al hilo. Por cada arista, verifica si se puede relajar
		 * (es decir, si se encuentra un camino más corto al nodo destino). Si es así, agrega una
		 * propuesta de actualización a la lista compartida. Los mensajes por consola permiten
		 * seguir el flujo de ejecución y depurar el comportamiento concurrente.
		 */
		@Override
		public void run() {
			System.out.println("[Hilo " + Thread.currentThread().getId() + "] Iniciando procesamiento de " + aristasAsignadas.size() + " aristas");
			for (Arista arista : aristasAsignadas) {
				int u = arista.origen;
				int v = arista.destino;
				int peso = arista.peso;

				// Lee la distancia actual al nodo origen de forma segura
				int distanciaU;
				synchronized (distancias) {
					distanciaU = distancias[u];
				}

				// Si se puede relajar la arista, propone una actualización
				if (distanciaU != Integer.MAX_VALUE && distanciaU + peso < distancias[v]) {
					System.out.println("[Hilo " + Thread.currentThread().getId() + "] Relajando arista " + u + " -> " + v + " con peso " + peso);
					// Agrega la actualización a la lista compartida de manera segura
					synchronized (actualizacionesCompartidas) {
						actualizacionesCompartidas.add(new Actualizacion(v, distanciaU + peso, u));
					}
				}
			}
			System.out.println("[Hilo " + Thread.currentThread().getId() + "] Finalizó procesamiento");
		}
	}

	// Aplica actualizaciones
	static boolean aplicarActualizaciones(List<Actualizacion> actualizaciones) {
		return false;
	}

	// Detecta ciclo negativo
	static boolean detectarCicloNegativo() {
		return false;
	}

	// Reconstruye la ruta
	static ArrayList<Integer> reconstruirRuta() {
		return null;
	}

	// Imprime la ruta
	static void imprimirRuta(ArrayList<Integer> ruta) {
	}

	// Guarda la salida en salidaThread.txt
	static void guardarSalida(ArrayList<Integer> ruta, int latencia) {
	}

	// Ejecuta una iteración de Bellman-Ford con hilos 
	static void bellmanFordThreads() {
		System.out.println("=== INICIO bellmanFordThreads ===");
		inicializarBellmanFord();
		ArrayList<ArrayList<Arista>> particiones = dividirAristas();

		// Lista compartida para actualizaciones
		List<Actualizacion> actualizacionesCompartidas = Collections.synchronizedList(new ArrayList<>());

		// Crear y lanzar los hilos
		List<Thread> hilos = new ArrayList<>();
		for (int i = 0; i < numHilos; i++) {
			Thread hilo = new HiloBellmanFord(particiones.get(i), actualizacionesCompartidas);
			hilos.add(hilo);
			hilo.start();
		}

		// Esperar a que todos los hilos terminen
		for (Thread hilo : hilos) {
			try {
				hilo.join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}

		// Mostrar actualizaciones propuestas
		System.out.println("Actualizaciones propuestas por los hilos:");
		for (Actualizacion act : actualizacionesCompartidas) {
			System.out.println("  Nodo: " + act.nodo + ", Distancia: " + act.distancia + ", Padre: " + act.padre);
		}
		System.out.println("=== FIN bellmanFordThreads ===");
	}

	// Main
	public static void main(String[] args) {
        
		if (args.length != 1) {
			System.out.println("archivo.txt");
			return;
		}

		leerArchivo(args[0]);
		bellmanFordThreads();
	}
}

File: LAB2_Ibaca_Lopez.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unistd.h>   // Para la creación y comunicación de procesos usando fork()
#include <sys/wait.h> // Para esperar a los procesos hijos
#include <chrono>     // Para medir tiempo
#include <climits>    // Constantes de límites enteros - Bellmand-Ford necesita "infinito"
#include <algorithm>

using namespace std;

struct Arista {
    int origen;
    int destino;
    int peso;
};

struct Actualizacion {
    int nodo;
    int distancia;
    int padre;
};

// ======================================================
// CONSTANTES
// ======================================================

const int INF = 1000000000;

// ======================================================
// VARIABLES GLOBALES
// ======================================================

int numProcesos;
int numNodos;
int nodoInicio;
int nodoFinal;

vector<Arista> aristas;
vector<int> distancias;
vector<int> predecesor;

// ======================================================
// FUNCIONES
// ======================================================

// lee el archivo de entrada y carga los datos del .txt
// tiene que leer numProcesos, numHilos, numNodos, nodoInicio, nodoFinal, origen, destino, peso...
void leerArchivo(string nombreArchivo) {

    ifstream archivo(nombreArchivo);
    
    if (!archivo.is_open()) {
        cerr << "Error al abrir el archivo de entrada: " << nombreArchivo << endl;
        exit(1);
    }

    int numHilos; // Solo se usa para saltar el dato de hilos (Java)
    archivo >> numProcesos;
    archivo >> numHilos; // Se ignora en C++
    archivo >> numNodos;
    archivo >> nodoInicio;
    archivo >> nodoFinal;

    aristas.clear();
    int origen, destino, peso;

    while (archivo >> origen >> destino >> peso) {
        aristas.push_back({origen, destino, peso});
    }

    archivo.close();
}

// inicializa las distancias y predecesores para Bellman-Ford
// tiene que preparar los vectores (distancias y predecesr) porque desde el nodo inicial -> si mismo (distancia es 0)
void inicializarBellmanFord(){
    distancias.assign(numNodos, INF); //crea el vector dejando todas las posiciones con infinito
    predecesor.assign(numNodos, -1); //padres con -1 porque no se sabe por donde se llega a cada nodo
    distancias[nodoInicio] = 0;
}

// divide las aristas entre los procesos hijos
// reparte las conexiones entre los hijos y retorna el vector donde cada pos es una lista de aristas de un hijo
vector<vector<Arista>> dividirAristas() {
    vector<vector<Arista>> grupos(numProcesos);  // cada posicion es un hioj

    for (int i = 0; i < (int)aristas.size(); i++) {
        int procesoAsignado = i % numProcesos;
        grupos[procesoAsignado].push_back(aristas[i]);
    }

    return grupos;
}
// crea un pipe para cada proceso hijo para la comunicación (pipes[i][2])
void crearPipes(vector<vector<int>>& pipes) {
    pipes.resize(numProcesos, vector<int>(2)); 

    for (int i = 0; i < numProcesos; i++) {
        if (pipe(pipes[i].data()) == -1) {
            cerr << "Error al crear pipe para el proceso " << i << endl;
            exit(1);
        }
    }
}

// aqui va todo lo que hace cada hijo
// cada hijo recibe aristas, distancias (copia) y lo del pipe
// recorre las aristas y revisa si es que puede mejorar alguna de las distancias
// cuando encuentra que una mejora, crea la actualizacion y lo manda al padre
void trabajoHijo(vector<Arista> aristasAsignadas, int pipeEscritura);

// es la función usada por el padre para leer las actualizaciones que mandó un hijo
vector<Actualizacion> leerPipe(int pipeLectura);

// padre hace las actualizaciones recibidas
// deberia retornar true si es que hay al menos un cambio
bool aplicarActualizaciones(vector<Actualizacion> actualizaciones);

// ve si es que hay algun ciclo negativo
// en caso de que haya significa que el camino no es válido
bool detectarCicloNegativo();

// usa predecesor[i] = algo para reconstruir la ruta
vector<int> reconstruirRuta();

// printea la ruta por pantalla
void imprimirRuta(vector<int> ruta);

// guarda la ruta y la latencia en salidaFork.txt
// ej: Ruta: 0 -> 1 -> 3 -> 11
//     latencia total: 75
void guardarSalida(vector<int> ruta, int latencia);

// ejecuta Bellman-Ford utilizando fork() y pipes
void bellmanFordFork();
// tiene que hacer en orden:
// 1. inicializarBellmanFord()
// 2. dividirAristas()
// 3. repetir numNodos - 1 veces: 
//   . crear hijos con fork()
//   . cada hijo analiza sus aristas
//   . cada hijo manda actualizaciones por pipe
//   . el padre lee los pipes
//   . padre aplica actualizaciones
//   . el padre espera a los hijos con wait()
// 4. detectar ciclo negativo
// 5. reconstruir ruta
// 6. imprimir resultado final
// 7. guardar salidaFork.txt

// ======================================================
// Main
// ======================================================

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cout << "Uso: ./programa archivo.txt" << endl;
        return 1;
    }

    leerArchivo(argv[1]);
    bellmanFordFork();
    return 0;
}